config/src/lib.rs

```diff
@@ -1,11 +1,296 @@
+    args_from_file.insert(0, cliargs.remove(0));
+    args_from_file.extend(cliargs);
+    // TODO: Convert to debug log statement
+    println!("final argv: {:?}", args_from_file);
+
+    // Return parsed config
+    Ok(new_matcher.get_matches_from(args_from_file))
 }
 
 #[cfg(test)]
 mod tests {
+    use std::ffi::OsString;
+
+    use crate::{get_matcher, ConfigDescription, ConfigOption, Configuration};
+    use std::env;
+
+    // Define a test configuration that can be used to run a few tests
+    struct TestConfig {}
+
+    impl TestConfig {
+        pub const TEST_PARAM: ConfigOption = ConfigOption {
+            name: "testparam",
+            default: "",
+            required: false,
+            takes_argument: true,
+            help: "Testhelp",
+            documentation: "Testdoc",
+        };
+        pub const TEST_PARAM2: ConfigOption = ConfigOption {
+            name: "testparam2",
+            default: "2",
+            required: false,
+            takes_argument: true,
+            help: "test2",
+            documentation: "test2",
+        };
+        pub const TEST_SWITCH: ConfigOption = ConfigOption {
+            name: "testswitch",
+            default: "",
+            required: false,
+            takes_argument: false,
+            help: "a switch that can be provided - or not",
+            documentation: "test doc switch",
+        };
+    }
+
+    impl ConfigDescription for TestConfig {
+        fn get_config(&self) -> Configuration {
+            Configuration {
+                name: "Test Tool",
+                version: "0.1",
+                about: "blabla",
+                options: vec![
+                    TestConfig::TEST_PARAM,
+                    TestConfig::TEST_PARAM2,
+                    TestConfig::TEST_SWITCH,
+                ],
+            }
+        }
+    }
+
+    #[test]
+    fn parse_single_param() {
+        let config = TestConfig {};
+        let command_line_args: Vec<OsString> = vec![
+            OsString::from("filename"),
+            OsString::from("--testparam"),
+            OsString::from("param1"),
+        ];
+        let matcher = get_matcher(&config, &"Test", command_line_args)
+            .expect("unexpected error occurred when parsing parameters");
+
+        assert!(matcher.is_present(TestConfig::TEST_PARAM.name));
+        assert_eq!(
+            matcher.value_of(TestConfig::TEST_PARAM.name).unwrap(),
+            "param1"
+        );
+    }
+
     #[test]
-    fn it_works() {
-        assert_eq!(2 + 2, 4);
+    fn parse_multiple_params() {
+        let config = TestConfig {};
+        let command_line_args: Vec<OsString> = vec![
+            OsString::from("filename"),
+            OsString::from("--testswitch"),
+            OsString::from("--testparam"),
+            OsString::from("param1"),
+            OsString::from("--testparam2"),
+            OsString::from("param2"),
+        ];
+        let matcher = get_matcher(&config, &"Test", command_line_args)
```

Comment:
Isn't "Test" already a &str? Does it need another "&" in front?

Reply:
yes, fixed